db/src/db.rs

```diff
@@ -759,12 +763,71 @@ impl DB {
+        }
+
+        // Like "UPDATE parts SET idx = CASE WHEN part = ? THEN ? WHEN part = ? THEN ? ELSE idx END".
+        let s = format!("UPDATE parts SET idx = CASE {} ELSE idx END",
+                        repeat("WHEN part = ? THEN ?").take(self.partition_map.len()).join(" "));
+
+        let params: Vec<&ToSql> = self.partition_map.iter().flat_map(|(name, partition)| {
+            once(name as &ToSql)
+                .chain(once(&partition.index as &ToSql))
+        }).collect();
+
+        // TODO: only cache the latest of these statements.  Changing the set of partitions isn't
+        // supported in the Clojure implementationat all, and might not be supported in Mentat soon,
+        // so this is very low priority.
+        let mut stmt = conn.prepare_cached(s.as_str())?;
+        stmt.execute(&params[..])
+            .map(|_c| ())
+            .chain_err(|| "Could not update partition map")
+    }
+
+    /// Allocate a single fresh entid in the given `partition`.
+    fn allocate_entid(&mut self, partition: String) -> i64 {
+        self.allocate_entids(partition, 1).start
+    }
+
+    /// Allocate `n` fresh entids in the given `partition`.
+    fn allocate_entids(&mut self, partition: String, n: usize) -> Range<i64> {
+        match self.partition_map.get_mut(&partition) {
+            Some(mut partition) => {
+                let idx = partition.index;
+                partition.index += n as i64;
+                idx..partition.index
+            },
+            // This is a programming error.
+            None => panic!("Cannot allocate entid from unknown partition: {}", partition),
+        }
+    }
+
+    /// Transact the given `entities` against the given SQLite `conn`, using the metadata in
+    /// `self.DB`.
+    ///
+    /// This approach is explained in https://github.com/mozilla/mentat/wiki/Transacting.
+    // TODO: move this to the transactor layer.
+    pub fn transact(&mut self, conn: &rusqlite::Connection, entities: &[Entity]) -> Result<TxReport> {
+        let tx_instant = now(); // Label the transaction with the timestamp when we first see it: leading edge.
+        let tx = self.allocate_entid(":db.part/tx".to_string());
+
+        // Eventually, this will be responsible for managing a SQLite transaction.  For now, it's
+        // just about the tx details.
+        self.transact_internal(conn, entities, tx, tx_instant)
+    }
+
     /// Transact the given `entities` against the given SQLite `conn`, using the metadata in
     /// `self.DB`.
     ///
     /// This approach is explained in https://github.com/mozilla/mentat/wiki/Transacting.
     // TODO: move this to the transactor layer.
-    pub fn transact_internal(&self, conn: &rusqlite::Connection, entities: &[Entity], tx_id: Entid, tx_instant: i64) -> Result<TxReport> {
+    pub fn transact_internal(&mut self, conn: &rusqlite::Connection, entities: &[Entity], tx_id: Entid, tx_instant: i64) -> Result<TxReport> {
```

Comment:
I see you commented further, but I think a duality in many Rust expressions:
- `&mut self -> ()` and caller owns coordination;
- `&self -> Self` and callee owns coordination.

I elected for the former since I expect to build a `Conn` to manage the coordination in #296.